content/articles/create-a-blockchain-in-python/index.md

```diff
@@ -0,0 +1,506 @@
+If the check returns True, then we’ve checked the proof and its valid, then we can assign it a value of true, else, we increment the new_proof by 1 which gives the miner another chance to try again and then we return the new proof.
+
+Your codebase should look this:
+```python
+def proof_of_work(self, previous_proof):
+   # miners proof submitted
+   new_proof = 1
+   # status of proof of work
+   check_proof = False
+   while check_proof is False:
+       # problem and algorithm based off the previous proof and new proof
+       hash_operation = hashlib.sha256(str(new_proof ** 2 - previous_proof ** 2).encode()).hexdigest()
+       # check miners solution to problem, by using miners proof in cryptographic encryption
+       # if miners proof results in 4 leading zero's in the hash operation, then:
+       if hash_operation[:4] == '0000':
+           check_proof = True
+       else:
+           # if miners solution is wrong, give mine another chance until correct
+           new_proof += 1
+   return new_proof
+
+```
+### Why Four Leading Zeros For Hash Operation?
+Now, why four(4) leading zeros? why not five(5) or six()6 or more, well the more leading zeros we require the more difficult it will be to mine a block, so, for the sake of the tutorial we have it at 4 to ensure we can mine a block faster.
+
+Also, the `hash_operation` algorithm we pass in must be non-symmetrical such that if the order of operation were to be reversed it wouldn’t result in the same value, e.g. a + b is equal to b+a, but a-b is not equal to b-a.
+
+Lastly, in our algorithm, we require both the new proof submitted and the previous proof from the existing block to effectively define an algorithm for miners to solve, that way the process is continuous and also linked to the existing blocks.
+
+This step is quite confusing but, if you read through the code and my comments, you will find it easy to grasp. Let’s proceed to the next step.
+
+### Generating a Hash
+In this step, we generate a hash for the entire block itself, so we create a `hash()` method that takes two parameters which are:
+- self
+- block.
+
+In the body of this function, we use JSON dumps to encode the block, and we return a cryptographic hash of the entire block.
+
+Your code should look like this:
+```python
+# generate a hash of an entire block
+def hash(self, block):
+   encoded_block = json.dumps(block, sort_keys=True).encode()
+   return hashlib.sha256(encoded_block).hexdigest()
+```
+
+### Checking The Validity of The Blockchain
+In this step, we create a function that checks if the entire blockchain is valid. This step is crucial in maintaining the integrity of our blockchain to ensure that none of our block is corrupt because if the data in a block changes, the hash changes and the link between each block in the chain is broken. This process is well explained in my previous [article](https://www.section.io/engineering-education/an-introduction-to-blockchain-architecture/#mining-and-proof-of-work).
```

Comment:
```suggestion
In this step, we create a function that checks if the entire blockchain is valid.

This step is crucial in maintaining the integrity of our blockchain to ensure that none of our blocks is corrupt because if the data in a block changes, the hash changes, and the link between each block in the chain is broken. This process is well explained in my previous [article](https://www.section.io/engineering-education/an-introduction-to-blockchain-architecture/#mining-and-proof-of-work).
```